### PR TITLE
MAC-26-View Employer account status

### PR DESCRIPTION
**Pull request by VasanthaKasirajan3008**

### CODE REVIEW

File: src/SFA.DAS.ProviderRegistrations/Application/Commands/SignedAgreementCommand/SignedAgreementCommandHandler.cs [hunk 1]
```diff
@@ -23,6 +23,10 @@ protected override async Task Handle(SignedAgreementCommand request, Cancellatio
             {
                 var invitation = await _db.Value.Invitations.SingleOrDefaultAsync(i => i.Reference == Guid.Parse(request.CorrelationId) && i.Status < (int) InvitationStatus.LegalAgreementSigned, cancellationToken);
                 invitation?.UpdateStatus((int) InvitationStatus.LegalAgreementSigned, DateTime.Now);
+
+                var invitationEvents = new InvitationEvent(invitation?.Id, (int)EventType.LegalAgreementSigned, DateTime.UtcNow);
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderRegistrations/Application/Commands/SignedAgreementCommand/SignedAgreementCommandHandler.cs:27]: I think we should be passing down the Datetime from the NServicebus event properties

File: src/SFA.DAS.ProviderRegistrations/Models/InvitationEvent.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+namespace SFA.DAS.ProviderRegistrations.Models
+{
+    public class InvitationEvent
+    {
+        public long Id { get; set; }
+
+        public long? InvitationId { get; set; }
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderRegistrations/Models/InvitationEvent.cs:9]: This class should include a navigation property for Invitation and vice versa

File: src/SFA.DAS.ProviderRegistrations/Application/Queries/GetInvitationEventByIdQuery/GetInvitationEventByIdQueryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+               .ToListAsync(cancellationToken);
+
+            if (invitationEvent.Count == 0)
+            {
+                return null;
+            }
+
+            var invitation = await _db.Value.Invitations
+                .Where(i => i.Id == request.InvitationId)
+                .SingleOrDefaultAsync(cancellationToken);
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderRegistrations/Application/Queries/GetInvitationEventByIdQuery/GetInvitationEventByIdQueryHandler.cs:39]: with navigation properties implemented as mentioned we should be able to do this in a single query

File: src/SFA.DAS.ProviderRegistrations/Application/Commands/UpsertUserCommand/UpsertUserCommandHandler.cs [hunk 4]
```diff
@@ -23,6 +23,10 @@ protected override async Task Handle(UpsertUserCommand request, CancellationToke
             {
                 var invitation = await _db.Value.Invitations.SingleOrDefaultAsync(i => i.Reference == Guid.Parse(request.CorrelationId) && i.Status < (int) InvitationStatus.AccountStarted, cancellationToken);
                 invitation?.UpdateStatus((int) InvitationStatus.AccountStarted, DateTime.Now);
+
+                var invitationEvents = new InvitationEvent(invitation?.Id, (int)EventType.AccountStarted ,DateTime.UtcNow);
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderRegistrations/Application/Commands/UpsertUserCommand/UpsertUserCommandHandler.cs:27]: once navigation properties have been implemented. Could do `invitation.InvitationEvents.Add(new InvitationEvent(...))`

**Review by cofaulco — COMMENTED** [review]
couple of comments

**Review by cofaulco — APPROVED** [review]

**Review by cofaulco — CHANGES_REQUESTED** [review]
couple of additional points

**Review by cofaulco — APPROVED** [review]

File: src/SFA.DAS.ProviderRegistrations/Application/Queries/GetInvitationEventByIdQuery/GetInvitationEventByIdQueryHandler.cs (hunk 3 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.ProviderRegistrations/Application/Queries/GetInvitationEventByIdQuery/GetInvitationEventByIdQueryHandler.cs]: done

File: src/SFA.DAS.ProviderRegistrations/Application/Commands/UpsertUserCommand/UpsertUserCommandHandler.cs (hunk 4 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.ProviderRegistrations/Application/Commands/UpsertUserCommand/UpsertUserCommandHandler.cs]: done

File: src/SFA.DAS.ProviderRegistrations.Database/Tables/InvitationEvents.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿CREATE TABLE [dbo].[InvitationEvents]
+(
+	[Id] BIGINT NOT NULL IDENTITY,
+	[InvitationId] BIGINT NOT NULL,    
+    [EventType] INT NOT NULL,
+    [Date] DATETIME NULL,
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.ProviderRegistrations.Database/Tables/InvitationEvents.sql]: Done

File: src/SFA.DAS.ProviderRegistrations.Web/Authentication/AuthenticationExtensions.cs [hunk 6]
```diff
@@ -13,14 +18,27 @@ public static class AuthenticationExtensions
     {
         public static IServiceCollection AddProviderIdamsAuthentication(this IServiceCollection services, IConfiguration config)
         {
-            var authenticationSettings = config.GetSection(ProviderRegistrationsConfigurationKeys.AuthenticationSettings).Get<AuthenticationSettings>();
+            if (config["UseAuthenictationStub"] != null && bool.Parse(config["UseAuthenictationStub"]))
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderRegistrations.Web/Authentication/AuthenticationExtensions.cs:21]: typo "Authentication"

File: src/SFA.DAS.ProviderRegistrations.Web/ViewModels/InvitationEventsViewModel.cs [hunk 7]
```diff
@@ -20,14 +22,64 @@ public class InvitationEventsViewModel
...
+                switch (Status)
+                {
+                    case InvitationStatusViewModel.PayeSchemeAdded:
+                    case InvitationStatusViewModel.LegalAgreementSigned:
+                    case InvitationStatusViewModel.InvitationComplete:
+                        return "Added";
+                }
+            }
+
+            return "PAYE scheme not added";
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderRegistrations.Web/ViewModels/InvitationEventsViewModel.cs:80]: are there unit tests for these newly calculated string properties?

**Review by cofaulco — CHANGES_REQUESTED** [review]
couple of comments

File: src/SFA.DAS.ProviderRegistrations.Web/Authentication/AuthenticationExtensions.cs (hunk 6 above)
Comment (hkf-tech) [reply src/SFA.DAS.ProviderRegistrations.Web/Authentication/AuthenticationExtensions.cs]: done

File: src/SFA.DAS.ProviderRegistrations.Web/ViewModels/InvitationEventsViewModel.cs (hunk 7 above)
Comment (hkf-tech) [reply src/SFA.DAS.ProviderRegistrations.Web/ViewModels/InvitationEventsViewModel.cs:80]: added some unit tests

**Review by cofaulco — APPROVED** [review]

**Review by cofaulco — APPROVED** [review]

**Pull request merged**